refactor: refactor x/amm module

kingcre opened the pull request:
## Description

- Simplify logic inside `AfterPoolOrdersExecuted`
- Store Pool's `CurrentSqrtPrice` instead of `CurrentPrice`
- Use `cremath.BigDec` for `CurrentSqrtPrice`

kingcre committed:
refactor: refactor AfterPoolOrdersExecuted

kingcre committed:
refactor: use CurrentSqrtPrice instead of CurrentPrice

kingcre committed:
refactor: refactor exchange module

- use sdk.Int for quantities
- change MaxPrice to 10^24
- store price as tick index for OrderBookOrderIndex
- apply ceiling/truncation per order, not per orderer

kingcre committed:
refactor: move MinOrderQty, MinOrderQuote into Market

also add MaxOrderQty, MaxOrderQuote

kingcre committed:
feat: add Market's FeeCollector

kingcre committed:
docs: update spec docs

also change MaxOrderQty, MaxOrderQuote to 10^30

kingcre committed:
test: fix tests

kingcre committed:
test: remove unused tests and add test case

kingcre committed:
refactor: change structure of params and apply code review suggestions

kingcre committed:
chore: apply code review suggestion

kingcre committed:
test: check EventOrderFilled

kingcre committed:
test: add test for MockOrderSource orders

kingcre committed:
merge 'refactor/exchange' into 'refactor/amm' and fix code

richard-bachman committed:
add: more test cases for batch and keeper

kingcre committed:
refactor: remove orderResultQueue

kingcre committed:
refactor: combine IterateTickInfosBelow/Above into IterateTickInfos

richard-bachman committed:
add: test for batch

richard-bachman committed:
add: test case for matching

kingcre committed:
feat: add cremath module which contains implementation of BigDec

kingcre committed:
refactor: make Pool to use BigDec

kingcre committed:
fix: temporarily limit the number of OrderSources that can be set

kingcre committed:
fix: apply code review suggestion

kingcre committed:
fix: apply code review suggestion

avoid store writes inside nested store iterations

kingcre committed:
fix: fix bug in AfterPoolOrdersExecuted and fix tests

use result.Paid + result.FeeReceived as paid amount

kingcre committed:
fix: emit EventRemoveLiquidity inside msg server

kingcre committed:
refactor: include FeePaid, FeeReceived in order result event

also rename Output to Received and add FeePaid, FeeReceived to
SwapRouteResult

kingcre committed:
fix: limit minOrderQty in swap and apply code review suggestions

kingcre committed:
Merge branch 'refactor/exchange' into refactor/amm

kingcre committed:
test: fix test

kingcre committed:
refactor: refactor code structure and add comments

kingcre committed:
test: fix test

kingcre committed:
test: fix test

richard-bachman commented on x/amm/types/math.go:
```diff
@@ -34,43 +35,43 @@ func LiquidityForAmounts(currentSqrtPrice, sqrtPriceA, sqrtPriceB sdk.Dec, amt0,
+	intermediate := sqrtPriceB.Sub(sqrtPriceA).MulIntMut(liquidity)
 	if roundUp {
-		return intermediate.QuoRoundUp(sqrtPriceB).QuoRoundUp(sqrtPriceA).Ceil().TruncateInt()
+		return intermediate.QuoRoundUpMut(sqrtPriceB).QuoRoundUpMut(sqrtPriceA).Ceil().TruncateInt()
```
What about using `intermediate.QuoRoundUpMut(sqrtPriceB.Mul(sqrtPriceA)).Ceil().TruncateInt()` instead of `intermediate.QuoRoundUpMut(sqrtPriceB).QuoRoundUpMut(sqrtPriceA).Ceil().TruncateInt()` in order to avoid using twice of round up?

richard-bachman commented on x/amm/types/math.go:
```diff
@@ -34,43 +35,43 @@ func LiquidityForAmounts(currentSqrtPrice, sqrtPriceA, sqrtPriceB sdk.Dec, amt0,
+		return intermediate.QuoRoundUpMut(sqrtPriceB).QuoRoundUpMut(sqrtPriceA).Ceil().TruncateInt()
 	}
-	return intermediate.QuoTruncate(sqrtPriceB).QuoTruncate(sqrtPriceA).TruncateInt()
+	return intermediate.QuoTruncateMut(sqrtPriceB).QuoTruncateMut(sqrtPriceA).TruncateInt()
```
Same comment as above.

kingcre committed:
build: bump up dependency versions

kingcre committed:
refactor: move NextOrderTick into types package

kingcre committed:
Merge pull request #220 from crescent-network/v5/richard-review2-refactor-exchange

Add test cases for batch and matching

kingcre committed:
Merge branch 'refactor/exchange' into refactor/amm

kingcre committed:
fix: make SqrtMut to be mutative

kingcre committed:
refactor: improve code readability

also
- rename types.RewardsPoolAddress to FarmingRewardsPoolAddress
  - also changed the address value

kingcre committed:
fix: change error messages and add tests

dudong2 commented on x/amm/keeper/pool_test.go:
```diff
@@ -236,12 +237,31 @@ func (s *KeeperTestSuite) TestPoolOrderMaxOrderPriceRatio() {
+	_, _, res := s.PlaceLimitOrder(
+		market.Id, ordererAddr, true, utils.ParseDec("5.1"), sdk.NewInt(20_000000), 0)
+
+	fmt.Println(res.IsMatched(), res.ExecutedQuantity)
```
```suggestion
	s.AssertEqual(res.IsMatched(), res.ExecutedQuantity)
```

kingcre committed:
test: fix test according to code review suggestion

kingcre committed:
refactor: rename simulation queries (Add/RemoveLiquidity)

kingcre committed:
fix: remove nested store iteration inside invariant check

kingcre committed:
fix: refactor code structure and apply code review suggestions

kingcre committed:
test: fix tests

kingcre committed:
test: fix test

after fixing bug in NextOrderTick, the result is changed

kingcre committed:
fix: set sqrt price, not price of the new pool

kingcre committed:
test: fix sim test and remove ineffectual assignment

kingcre committed:
test: fix sim test

kingcre committed:
fix: apply code review suggestions

kingcre committed:
fix: cap order price limit within [min price, max price]

kingcre committed:
fix: do not use reserveBalances.Min in RemoveLiquidity

kingcre committed:
fix: change error message from ValidatePriceRange and add tests

kingcre committed:
feat: add CanRemoveLiquidityInvariant

kingcre committed:
refactor: calculate next price based on amt0 (wip)

also calculate order qty based on amt0 delta too

dudong2 commented on x/amm/keeper/position_test.go:
```diff
@@ -198,3 +383,23 @@ func (s *KeeperTestSuite) TestRemoveSmallLiquidity() {
+
+	lpAddr := s.FundedAccount(1, enoughCoins)
+	_, _, _, err := s.keeper.AddLiquidity(
+		s.Ctx, lpAddr, lpAddr, pool.Id, sdk.NewDecWithPrec(1, sdk.Precision), utils.ParseDec("10"),
```
```suggestion
		s.Ctx, lpAddr, lpAddr, pool.Id, types.MinPrice, utils.ParseDec("10"),
```

kingcre replied:
This is intended to test if `AddLiquidity` properly rejects invalid lower/upper price.

dudong2 commented on x/amm/keeper/position_test.go:
```diff
@@ -198,3 +383,23 @@ func (s *KeeperTestSuite) TestRemoveSmallLiquidity() {
+			"0.000000000000000001 < 0.000000000000010000: invalid request")
+
+	_, _, _, err = s.keeper.AddLiquidity(
+		s.Ctx, lpAddr, lpAddr, pool.Id, utils.ParseDec("4"), sdk.NewDecFromInt(sdk.NewIntWithDecimal(1, 40)),
```
```suggestion
		s.Ctx, lpAddr, lpAddr, pool.Id, utils.ParseDec("4"), types.MaxPrice),
```

kingcre replied:
Ditto

kingcre committed:
fix: negative extraQuote is allowed up to -(numOrders)

kingcre committed:
fix: min qty/quote must be positive

kingcre committed:
test: fix tests

kingcre committed:
refactor: introduce new Ledger struct to keep track of coin diff

kingcre committed:
fix: move quote dust to the OrderSource orderer(amm)

kingcre committed:
fix: fix NextOrderTick to reflect changes in pool order logic

kingcre committed:
Merge branch 'refactor/exchange' into refactor/amm

kingcre committed:
fix: fix bug in MatchState.Result and quote dust calculation

kingcre committed:
Merge branch 'refactor/exchange' into refactor/amm

kingcre committed:
test: fix tests

kingcre committed:
chore: remove unused variable

kingcre commented:
Closing. See #223
